### PR TITLE
CP-50354 Check IP

### PR DESCRIPTION
**Pull request by acefei**
Tested with ssh connection
```
Plan: 1 to add, 0 to change, 1 to destroy.
xenserver_vm.linux_vm: Destroying... [id=5a16fbde-4253-d8ab-c613-c51153108b4c]
xenserver_vm.linux_vm: Destruction complete after 8s
xenserver_vm.linux_vm: Creating...
xenserver_vm.linux_vm: Still creating... [10s elapsed]
xenserver_vm.linux_vm: Still creating... [20s elapsed]
xenserver_vm.linux_vm: Still creating... [30s elapsed]
xenserver_vm.linux_vm: Still creating... [40s elapsed]
xenserver_vm.linux_vm: Provisioning with 'remote-exec'...
xenserver_vm.linux_vm (remote-exec): Connecting to remote host via SSH...
xenserver_vm.linux_vm (remote-exec):   Host: 10.71.153.62
xenserver_vm.linux_vm (remote-exec):   User: root
xenserver_vm.linux_vm (remote-exec):   Password: true
xenserver_vm.linux_vm (remote-exec):   Private key: false
xenserver_vm.linux_vm (remote-exec):   Certificate: false
xenserver_vm.linux_vm (remote-exec):   SSH Agent: false
xenserver_vm.linux_vm (remote-exec):   Checking Host Key: false
xenserver_vm.linux_vm (remote-exec):   Target Platform: unix
xenserver_vm.linux_vm (remote-exec): Connected!
xenserver_vm.linux_vm (remote-exec): PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
xenserver_vm.linux_vm (remote-exec): NAME="Debian GNU/Linux"
xenserver_vm.linux_vm (remote-exec): VERSION_ID="12"
xenserver_vm.linux_vm (remote-exec): VERSION="12 (bookworm)"
xenserver_vm.linux_vm (remote-exec): VERSION_CODENAME=bookworm
xenserver_vm.linux_vm (remote-exec): ID=debian
xenserver_vm.linux_vm (remote-exec): HOME_URL="https://www.debian.org/"
xenserver_vm.linux_vm (remote-exec): SUPPORT_URL="https://www.debian.org/support"
xenserver_vm.linux_vm (remote-exec): BUG_REPORT_URL="https://bugs.debian.org/"
xenserver_vm.linux_vm: Creation complete after 49s [id=f282c82a-43fc-e966-e440-6e2d9cf5a1ba]

Apply complete! Resources: 1 added, 0 changed, 1 destroyed.
make[1]: Leaving directory '/home/feis/Documents/terraform-provider-xenserver'
```

Test Acc
```source .env && TF_ACC=1 go test ./xenserver/ -v   -timeout 120m
=== RUN   TestAccNetworkDataSource
--- PASS: TestAccNetworkDataSource (0.78s)
=== RUN   TestAccVlanResource
--- PASS: TestAccVlanResource (3.76s)
=== RUN   TestAccNICDataSource
--- PASS: TestAccNICDataSource (0.79s)
=== RUN   TestAccPifDataSource
--- PASS: TestAccPifDataSource (0.74s)
=== RUN   TestAccSnapshotResource
--- PASS: TestAccSnapshotResource (26.22s)
=== RUN   TestAccSRDataSource
--- PASS: TestAccSRDataSource (0.83s)
=== RUN   TestAccNFSResource
--- PASS: TestAccNFSResource (9.14s)
=== RUN   TestAccSRResourceLocal
--- PASS: TestAccSRResourceLocal (4.74s)
=== RUN   TestAccSRResourceShared
--- PASS: TestAccSRResourceShared (8.47s)
=== RUN   TestAccVDIResource
--- PASS: TestAccVDIResource (19.85s)
=== RUN   TestAccVMDataSource
--- PASS: TestAccVMDataSource (12.37s)
=== RUN   TestAccVMResource
--- PASS: TestAccVMResource (44.37s)
=== RUN   TestAccLinuxVMResource
--- PASS: TestAccLinuxVMResource (7.37s)
PASS
ok      terraform-provider-xenserver/xenserver  139.442s```

### CODE REVIEW

File: xenserver/vm_utils.go [hunk 1]
```diff
@@ -943,16 +978,116 @@ func setVMResourceModel(ctx context.Context, session *xenapi.Session, vmRef xena
...
+	switch {
+	case ip.IsLinkLocalMulticast():
+		fallthrough
+	case ip.IsLinkLocalUnicast():
+		fallthrough
+	case ip.IsLoopback():
+		fallthrough
+	case ip.IsMulticast():
+		return false
+	default:
```
Comment (xueqingz) [inline xenserver/vm_utils.go:998]: default null?

File: xenserver/vm_utils.go [hunk 2]
```diff
@@ -943,16 +978,116 @@ func setVMResourceModel(ctx context.Context, session *xenapi.Session, vmRef xena
...
+	}
+
+	// set timeout channel
+	timeoutChan := time.After(time.Duration(checkIPTimeout) * time.Second)
+	for {
+		select {
+		case <-timeoutChan:
+			// raise error when user wants to check IP address but the IP address is not found
+			if vmRecord.OtherConfig["tf_check_ip_timeout"] != "0" {
+				return "", errors.New("get IP address from VM timeout")
```
Comment (xueqingz) [inline xenserver/vm_utils.go:1037]: What does error "get IP address from VM timeout" means?

File: xenserver/vm_utils.go [hunk 3]
```diff
@@ -943,16 +978,116 @@ func setVMResourceModel(ctx context.Context, session *xenapi.Session, vmRef xena
...
+		if err != nil {
+			return errors.New(err.Error())
+		}
+	}
+
 	return nil
 }
 
+func checkIP(ctx context.Context, session *xenapi.Session, vmRecord xenapi.VMRecord) (string, error) {
+	checkIPTimeout, err := strconv.Atoi(vmRecord.OtherConfig["tf_check_ip_timeout"])
```
Comment (xueqingz) [inline xenserver/vm_utils.go:1025]: Could we add a check for checkIPTimeout  ,
if checkIPTimeout <1 . return "", nil

File: xenserver/vm_utils.go (hunk 2 above)
Comment (acefei) [reply xenserver/vm_utils.go:1037]: get ip timeout

File: xenserver/vm_utils.go [hunk 4]
```diff
@@ -943,16 +978,116 @@ func setVMResourceModel(ctx context.Context, session *xenapi.Session, vmRef xena
...
+	var ip = ""
+	vmGuestMetricsRef := vmRecord.GuestMetrics
+	vmGuestMetricRecord, err := xenapi.VMGuestMetrics.GetRecord(session, vmGuestMetricsRef)
+	if err != nil {
+		return ip, errors.New(err.Error())
+	}
+
+	for k, v := range vmGuestMetricRecord.Networks {
+		tflog.Debug(ctx, "-----> getIPAddressFromMetrics key: "+k+" value: "+v)
+		if strings.HasSuffix(k, "ip") {
```
Comment (xueqingz) [inline xenserver/vm_utils.go:1061]: Could you give some examples here about  the key and value to help understand? what the other key looks like? and what the key we want looks like?

File: xenserver/vm_utils.go [hunk 5]
```diff
@@ -943,16 +978,116 @@ func setVMResourceModel(ctx context.Context, session *xenapi.Session, vmRef xena
...
+			if isValidIpAddress(net.ParseIP(v)) {
+				ip = v
+			}
```
Comment (xueqingz) [inline xenserver/vm_utils.go:1064]: Do we need break here?

File: xenserver/vm_utils.go [hunk 6]
```diff
@@ -943,16 +978,116 @@ func setVMResourceModel(ctx context.Context, session *xenapi.Session, vmRef xena
...
+		}
+	}
+
+	// raise error when user wants to check IP address but the IP address is not found
+	if vmRecord.OtherConfig["tf_check_ip_timeout"] != "0" && ip == "" {
```
Comment (xueqingz) [inline xenserver/vm_utils.go:1069]: Do we need to consider vmRecord.OtherConfig["tf_check_ip_timeout"] != "0"?
or just 
```
if ip == "" {
	return ip, errors.New("unable to get IP address from VM")
}

```

File: examples/resources/xenserver_vm/resource.tf [hunk 7]
```diff
@@ -58,6 +59,33 @@ resource "xenserver_vm" "vm" {
   }
 }
 
-output "vm_out" {
-  value = xenserver_vm.vm
+# Create a Linux VM that is cloned from the custom template
+resource "xenserver_vm" "linux_vm" {
```
Comment (xueqingz) [inline examples/resources/xenserver_vm/resource.tf:63]: Do you miss check_ip_timeout here?

File: examples/resources/xenserver_vm/resource.tf [hunk 8]
```diff
@@ -58,6 +59,33 @@ resource "xenserver_vm" "vm" {
...
+  name_label     = "Linux VM"
+  template_name  = "CustomTemplate"
+  static_mem_max = 4 * 1024 * 1024 * 1024
+  vcpus          = 4
+
+  # Don't need to set up a hard drive if the custom template includes it
+
+  # Need to set up the network interface if the default one in the custom template would be removed
```
Comment (xueqingz) [inline examples/resources/xenserver_vm/resource.tf:71]: In current code , TF always remove the network interface in custom template, not if.

File: examples/resources/xenserver_vm/resource.tf [hunk 9]
```diff
@@ -58,6 +59,33 @@ resource "xenserver_vm" "vm" {
...
+    {
+      network_uuid = data.xenserver_network.network.data_items[0].uuid,
+      device       = "0"
+    },
+  ]
+
+  connection {
+    type     = "ssh"
+    user     = "root"
+    password = "******"
```
Comment (xueqingz) [inline examples/resources/xenserver_vm/resource.tf:82]: for this user name and password, are them defined in Template?

File: xenserver/vm_utils.go (hunk 3 above)
Comment (acefei) [reply xenserver/vm_utils.go:1025]: There is a validator to limit number greater than 0 and the following code would check if it is equals "0", so here don't need to extra checking.

File: examples/resources/xenserver_vm/resource.tf (hunk 9 above)
Comment (acefei) [reply examples/resources/xenserver_vm/resource.tf:82]: No, they are user-defined

File: xenserver/vm_utils.go (hunk 4 above)
Comment (acefei) [reply xenserver/vm_utils.go:1061]: we can found the key-value by xe vm-param-list, eg:
```
networks (MRO): 0/ip: 10.62.66.9; 0/ipv4/0: 10.62.66.9; 0/ipv6/1: fe80::34d4:3eff:fee8:9a0e; 0/ipv6/0: fd06:7768:b9e5:8c30:34d4:3eff:fee8:9a0e
```

File: xenserver/vm_utils.go (hunk 5 above)
Comment (acefei) [reply xenserver/vm_utils.go:1064]: Typecally, there is only one key with ip suffix, but break is more efficient.

File: xenserver/vm_utils.go (hunk 6 above)
Comment (acefei) [reply xenserver/vm_utils.go:1069]: No, there are two scenarios:
1. check_ip_timeout not set or set to 0, the ip is always return "", this is expected, won't raise error.
2. user set check_ip_timeout, in this way, raise err if ip == "".

**Review by xueqingz — APPROVED** [review]

**Review by xihuan-citrix — APPROVED** [review]

**Pull request merged**